daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -84,147 +50,168 @@ private[engine] final class ValueTranslator(compiledPackages: CompiledPackages)
+    def go(
+        ty: Type,
+        value: Value[ContractId],
+        subst: ClosedSubstitution[Type] = ClosedSubstitution.empty,
+        nesting: Int = 0,
+    ): SValue =
       if (nesting > Value.MAXIMUM_NESTING) {
         fail(s"Provided value exceeds maximum nesting level of ${Value.MAXIMUM_NESTING}")
       } else {
-        val newNesting = nesting + 1
-        (ty, value) match {
-          // simple values
-          case (TUnit, ValueUnit) =>
-            SValue.SUnit
-          case (TBool, ValueBool(b)) =>
-            if (b) SValue.SValue.True else SValue.SValue.False
-          case (TInt64, ValueInt64(i)) =>
-            SValue.SInt64(i)
-          case (TTimestamp, ValueTimestamp(t)) =>
-            SValue.STimestamp(t)
-          case (TDate, ValueDate(t)) =>
-            SValue.SDate(t)
-          case (TText, ValueText(t)) =>
-            SValue.SText(t)
-          case (TNumeric(TNat(s)), ValueNumeric(d)) =>
-            Numeric.fromBigDecimal(s, d).fold(fail, SValue.SNumeric)
-          case (TParty, ValueParty(p)) =>
-            SValue.SParty(p)
-          case (TContractId(_), ValueContractId(c)) =>
-            cids += c
-            SValue.SContractId(c)
-
-          // optional
-          case (TOptional(elemType), ValueOptional(mb)) =>
-            SValue.SOptional(mb.map((value: Value[ContractId]) => go(elemType, value, newNesting)))
-
-          // list
-          case (TList(elemType), ValueList(ls)) =>
-            SValue.SList(ls.map((value: Value[ContractId]) => go(elemType, value, newNesting)))
-
-          // textMap
-          case (TTextMap(elemType), ValueTextMap(entries)) =>
-            SValue.SGenMap(
-              isTextMap = true,
-              entries = entries.iterator.map {
-                case (k, v) => SValue.SText(k) -> go(elemType, v, newNesting)
-              }
-            )
-
-          // genMap
-          case (TGenMap(keyType, valueType), ValueGenMap(entries)) =>
-            SValue.SGenMap(
-              isTextMap = false,
-              entries = entries.iterator.map {
-                case (k, v) => go(keyType, k, newNesting) -> go(valueType, v, newNesting)
-              }
-            )
-
-          // variants
-          case (TTyConApp(typeVariantId, tyConArgs), ValueVariant(mbId, constructorName, val0)) =>
-            mbId.foreach(id =>
-              if (id != typeVariantId)
-                fail(
-                  s"Mismatching variant id, the type tells us $typeVariantId, but the value tells us $id"))
-            val pkg = unsafeGetPackage(typeVariantId.packageId)
-            val (dataTypParams, variantDef) =
-              assertRight(SignatureLookup.lookupVariant(pkg, typeVariantId.qualifiedName))
-            variantDef.constructorRank.get(constructorName) match {
+        ty match {
+          case TVar(name) =>
+            subst.get(name) match {
+              case Some((newSubst, newTyp)) =>
+                go(newTyp, value, newSubst, nesting)
```

Comment:
This isn't really lazy, is it? The value in the Substitution will be substituted n times, if it appears n times? To be lazy, you would memoize the values in the substitution, so you only need to substitute at most once.

Reply:
Your are right, "lazy" is probably not the good terminology.  
The point of this PR is to avoid the "deep" substitution that was done by `replaceParameters` before.

Reply:
Do you have a better idea, on how to reformulate the PR description and the comment ?

Reply:
Implementing a real lazy version seems quite more involved.

Reply:
I'm not sure what the substitution is doing... So I don't think I could describe it.

Reply:
I think I’ve seen this called delayed substitution.

Reply:
Going back to the performance question:

Previously we did eager substitution. This means we have to traverse the whole type every time we substitute which is bad.

Now we perform delayed substitution. We only substitute when we encounter the variable. However I think there is a problem here as @sofiafaro-da pointed out. This substitution step isn’t a constant-time operation, you are doing actual work here since the type you are substituting for the variable may need more substitutions applied to it and you’ll repeat that work for every occurence of the variable. I believe we didn’t suffer from that repeated work before.

One way to avoid this would be to normalize types when we insert them in the substitution map. This is still better than what we did before and it turns substitution into a constant time operation since you don’t have to recurse on the term you substitute. It does a bit of unnecessary work in case the type isn’t used at all but that seems fine and afaict it is strictly better than what we do atm.